src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs

```diff
@@ -4849,17 +4849,584 @@ await foreach (var o in new Enumerable())
             CompileAndVerify(comp, expectedOutput: "StructAwaitable1StructAwaitable2");
         }
 
+        [Fact]
+        public void TestInvalidForeachOnConstantNullObject()
+        {
+            var source = @"
+using System;
+using System.Threading.Tasks;
+public class C
+{
+    public static async Task Main()
+    {
+        await foreach (var i in (object)null)
+        {
+            Console.Write(i);
+        }
+    }
+}";
+            CreateCompilation(source, parseOptions: TestOptions.RegularPreview)
+                .VerifyDiagnostics(
+                    // (8,33): error CS8411: Asynchronous foreach statement cannot operate on variables of type 'object' because 'object' does not contain a suitable public instance or extension definition for 'GetAsyncEnumerator'
+                    //         await foreach (var i in (object)null)
+                    Diagnostic(ErrorCode.ERR_AwaitForEachMissingMember, "(object)null").WithArguments("object", "GetAsyncEnumerator").WithLocation(8, 33)
+                    );
+        }
+
+        [Fact]
+        public void TestConstantNullObjectImplementingIEnumerable()
+        {
+            var source = @"
+using System;
+using System.Threading.Tasks;
+using System.Collections.Generic;
+public class C
+{
+    public static async Task Main()
+    {
+        await foreach (var i in (IAsyncEnumerable<int>)null)
+        {
+            Console.Write(i);
+        }
+    }
+}";
+            CreateCompilationWithTasksExtensions(new[] { source, AsyncStreamsTypes }, parseOptions: TestOptions.RegularPreview)
+                .VerifyDiagnostics(
+                    // (9,33): error CS0186: Use of null is not valid in this context
```

Comment:
>CS0186 [](start = 37, length = 6)

This error makes no sense to me. The "context" of this null is the target of a cast expression. Since it is being cast to a class type, it is valid. Looking at [https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/async-streams.md](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/async-streams.md) I find nothing to support this error.

Reply:
Recorded as [https://github.com/dotnet/roslyn/issues/45616](https://github.com/dotnet/roslyn/issues/45616)

---
In reply to: [449174471](https://github.com/dotnet/roslyn/pull/45160#discussion_r449174471) [](ancestors = 449174471)

Reply:
Are you saying the current behavior is a bug?

The comment here says this isn't to spec but is the same as dev10:

https://github.com/dotnet/roslyn/blob/be56bf44b7bcd2462c123aca88282791b4e5bca0/src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs#L679-L684

Given that this will always throw, I don't know if it's worth fixing this behaviour.